### ISSUE TITLE
Weird bind implementation?

### ISSUE THREAD

**Issue by deliciouslytyped** [python-parsy/parsy#33]
Hi, thanks for making parsy, I like it!

I'm currently succeeding at shooting myself in the foot with heterogeneous inputs, because I want to be able to process a list of "symbols" and strings:
```python
class Indent():
  def __init__(self, depth):
    self.depth=depth
  def __repr__(self):
    return "Indent(%s)" % self.depth

newline = string("\r\n")
preline_ = (newline.should_fail("not newline") >> any_char).many().concat()
line = preline_ << newline
take = test_item(lambda x: True, "any nested")

```

I'm probably not understanding something, why do I need to use `.bind` like this?:
```python
>>> wrap = lambda fn: lambda res: success(fn.parse(res))
>>> ( sideToken >> take.bind(wrap(line)) ).many()
        .parse([Indent(0), "asd\r\n", Indent(1), "asdf\r\n"])
['asd', 'asdf']
```
instead of: (the following is pseudocode)
```python
>>> ( sideToken >> take.bind(line) ).many() \
        .parse([Indent(0), "asd\r\n", Indent(1), "asdf\r\n"])
['asd', 'asdf']
```

Which is to say, why do I have to add an extra layer of wrapping and calling `.parse`?

**Comment by deliciouslytyped** [comment]
Related: I seem to have to define wrap something like the following to get backtracking to work?
```
def wrap(fn):
  def thing(res):
    try:
      result = fn.parse(res)
      return success(result)
    except:
      return fail("idk")
  return thing
```

What's the correct way to do this?

**Comment by spookylukey** [comment]
@jneen do you have any input on this? I've never actually used `bind` directly, perhaps you could answer this question more easily than me?

**Comment by deliciouslytyped** [comment]
Ok, this makes more sense when used like in https://github.com/python-parsy/parsy/issues/36#issuecomment-564736240 , I need to revisit this later.

**Comment by jneen** [comment]
I'm not sure what you're trying to do, could you explain in a little more detail? The "pseudocode" example you gave passes a `Parser` instance to `.bind` - but `.bind` takes a function that returns a parser. It's a little hard to follow because your variable named `fn` is not actually a function, but a parser.

**Comment by jneen** [comment]
Oh, I think I see what's happening. You are mixing strings and tokens, and want to "re-parse" the parse result when it's not a token. I... would suggest just tokenizing your entire input before passing it to parsy if that's the direction you want to go.

**Comment by jneen** [comment]
If you are really dedicated to this approach you might want to define something like:

``` python
def reparse(parser):
    test_item(lambda x: isinstance(x, unicode), 'not a string').map(parser.parse)
```

**Comment by jneen** [comment]
@spookylukey for reference, `bind` is just a way to dynamically determine the next parser, which you can naturally do with `@generate` functions. These are equivalent:

```
parser = first.bind(second)

@generate
def parser():
    res = yield first
    return (yield second(res))
```
